akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -0,0 +1,501 @@
+
+  final class InputState(signalDemand: AsyncCallback[Long]) {
+    private var untilNextDemandSignal = DemandThreshold
+
+    def onElement(): Unit = {
+      untilNextDemandSignal -= 1
+      if (untilNextDemandSignal == 0) {
+        untilNextDemandSignal = DemandThreshold
+        signalDemand.invoke(DemandThreshold)
+      }
+    }
+
+    def close(): Unit = signalDemand.invoke(MergeHub.Cancel)
+
+  }
+
+  final class MergedSourceLogic(_shape: Shape, producerCount: AtomicLong) extends GraphStageLogic(_shape) with OutHandler {
+    private val queue = new AbstractNodeQueue[Event] {}
+    @volatile private[this] var needWakeup = false
+    @volatile private[this] var shuttingDown = false
+
+    private[this] val demands = new java.util.HashMap[Long, InputState]
+    private[this] val wakeupCallback = getAsyncCallback[NotUsed]((_) ⇒
+      // We are only allowed to dequeue if we are not backpressured. See comment in tryProcessNext() for details.
+      if (isAvailable(out)) tryProcessNext(firstAttempt = true)
+    )
+
+    setHandler(out, this)
+
+    // Returns true when we have not consumed demand, false otherwise
+    private def onEvent(ev: Event): Boolean = ev match {
+      case Element(id, elem) ⇒
+        demands.get(id).onElement()
+        push(out, elem)
+        false
+      case Register(id, callback) ⇒
+        // TODO: Check for already existing state? Should not happen if works correctly though
+        demands.put(id, new InputState(callback))
+        true
+      case Deregister(id) ⇒
+        demands.remove(id)
+        true
+    }
+
+    override def onPull(): Unit = tryProcessNext(firstAttempt = true)
+
+    @tailrec private def tryProcessNext(firstAttempt: Boolean): Unit = {
+      val nextElem = queue.poll()
+
+      // That we dequeue elements from the queue when there is demand means that Register and Deregister messages
+      // might be delayed for arbitrary long. This is not a problem as Register is only interesting if it is followed
+      // by actual elements, which would be delayed anyway by the backpressure.
+      // Unregister is only used to keep the map growing too large, but otherwise it is not critical to process it
+      // timely. In fact, the only way the map could keep growing would mean that we dequeue Registers from the
+      // queue, but then we will eventually reach the Deregister message, too.
+      if (nextElem ne null) {
+        needWakeup = false
+        if (onEvent(nextElem)) tryProcessNext(firstAttempt = true)
+      } else {
+        needWakeup = true
+        // additional poll() to grab any elements that might missed the needWakeup
+        // and have been enqueued just after it
+        if (firstAttempt)
+          tryProcessNext(firstAttempt = false)
+      }
+    }
+
+    def isShuttingDown: Boolean = shuttingDown
+
+    // External API
+    def enqueue(ev: Event): Unit = {
+      queue.add(ev)
+      if (needWakeup) {
```

Comment:
Not exactly :)

So, the reason here for using a dedicated queue is performance. By going through this queue most of the time I can avoid the cost of an AsyncCallback, which would go through the stream actor, costing an Envelope, the actor mailbox processing logic, dispatch through receive, then dispatch to the interpreter. Also, this would happen for each message making it hard for the interpreter to warm up. 

The above design relies on that when we read an element, we push it downstream, and only ask for the next one when a pull comes back. At that time there is high likelihood that there is already an element waiting, so we just take it immediately directly from the queue. Basically the execution turns into a poll-work loop. The nice thing here is that the interpreter after a while suspends the execution by sending an actor self-message, so there is no danger of blocking a thread for too long.

As for why this implementation is different, the answer is because the requirements are different. In the actor it is harmful to accidentally do a double-wakeup and submit it to the thread-pool twice. We don't have this issue here, because our wakeup here simply piggy-backs on the actor and uses a normal AsyncCallback (an actor message) to handle this. I.e. we only need to solve one half of the problem, the rest is handled by the actor already. In this code, receiving two wakeups is not harmful the only issue would be missing wakeups as it leads to deadlock.

In summary, this side-queue is basically used to amortize the cost of going through the full actor message path (through AsyncCallback) by doing it only once in a while.

As for high-intensity testing, this exact logic is used in the Artery SendQueue, and Artery benchmarks are running every night. If we find a bug there, we can immediately fix it here, too. Anyway, I probably go through the logic here again, and try to make a proper proof why it works and add it as a comment.